docs/usage/PRODUCT_SYNC.md

```diff
@@ -59,7 +60,130 @@ order for the sync to resolve the actual ids of those references, those `key`s h
 ````
-[More information about Sync Options](SYNC_OPTIONS.md). 
+
+#### About SyncOptions
+`SyncOptions` is an object which provides a place for users to add certain configurations to customize the sync process.
+Here are configurations included :
+
+##### 1. `errorCallback`
+A callback that is called whenever an error event occurs during the sync process. It contains the following
+information about the error-event:
+
+* sync exception
+* product draft from the source
+* product of the target project
+* the update-actions, which failed
+
+##### Example 
+````java
+ final Logger logger = LoggerFactory.getLogger(MySync.class);
+ final ProductSyncOptions productsyncOptions = ProductSyncOptionsBuilder
+         .of(sphereClient)
+         .errorCallback((syncException, draft, product, updateActions) -> 
+            logger.error(syncException.getMessage(), syncException)).build();
+````
+    
+##### 2. `warningCallback`
+A callback that is called whenever a warning event occurs during the sync process. It contains the following 
+information about the warning message:
+
+* sync exception
+* product draft from the source 
+* product of the target project
+
+##### Example 
+````java
+ final Logger logger = LoggerFactory.getLogger(MySync.class);
+ final ProductSyncOptions productsyncOptions = ProductSyncOptionsBuilder
+         .of(sphereClient)
+         .warningCallback((syncException, draft, product, updateActions) -> 
+            logger.warn(syncException.getMessage(), syncException)).build();
+````
+
+##### 3. `beforeUpdateCallback`
+During the sync process if a target product and a product draft are matched, this callback can be used to intercept 
+the **_update_** request just before it is sent to CTP. It contains following information :
+ 
+ * product draft from the source
+ * product from the target project
+ * update actions that were calculated after comparing both.  
+
+##### Example
+````java
+final TriFunction<
+        List<UpdateAction<Product>>, ProductDraft, Product, List<UpdateAction<Product>>> beforeUpdateProductCallback =
+            (updateActions, newProductDraft, oldProduct) ->  updateActions.stream()
+                    .filter(updateAction -> !updateAction.getAction().isEmpty())
+                    .collect(Collectors.toList());
+                        
+final ProductSyncOptions productSyncOptions = 
+        ProductSyncOptionsBuilder.of(sphereClient).beforeUpdateCallback(beforeUpdateCallback).build();
+````
+
+##### 4. `beforeCreateCallback`
+During the sync process if a product draft should be created, this callback can be used to intercept 
+the **_create_** request just before it sent to CTP.  It contains following information : 
+
+ * product draft that should be created.
+ 
+##### Example
+````java
+final Function<ProductDraft, ProductDraft> beforeCreateProductCallback =
+             (callbackDraft) -> {
+                 final String newProductDraftKey = format("%s%s", callbackDraft.getKey(), "NEW_PROJECT");
```

Comment:
Replacing the key might be not a useful example as we match by key before we call this callback and if the new key is already used in the project then it will run into API error. There are not many useful examples for this callback. One could argue if it is needed at all because why use before create callback if you can modify your drafts before they are sent to the library? Following examples might be slightly more practical from business perspective: based on draft data populate `searchKeywords` or generate `slug` or set `published` flag to `true`.

Analog to `beforeUpdateCallback` you could create a class with `beforeCreateCallback` and link it here.

Reply:
Change it to adapt `searchKeywords` logging as example.
https://github.com/commercetools/commercetools-sync-java/pull/586/commits/4c576551f73acca4ccca9538f31e269b306cb1f8

Reply:
Logging search keywords is not really useful example as I meant rather their generation based on the draft data (concatenation of product name + brand and some other variant attribute - but this might be to long example). 

Following case seems to be more realistic/useful:
Set `publish` flag to `true` if category reference is set to the product draft. Because if it is set then category reference resolution was successful. One may want to publish new product if there are assigned to the at least one category.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/586/commits/b8fd75a22ba4e3f2e341637ab39eae9439575154